### DISCUSSION TITLE
Swiper with Alpine JS: Uncaught ReferenceError: slide is not defined

### DISCUSSION THREAD

**Discussion by huubl** [alpinejs/alpine#850] [1. Help] [answered]
Hi,

I try to use [Swiper](https://swiperjs.com) with Alpine JS. It seem to be working, but somehow I get an error:

```
VM23631:3 Uncaught ReferenceError: slide is not defined
    at eval (eval at saferEval (alpine.js:118), <anonymous>:3:36)
    at saferEval (alpine.js:118)
    at Component.evaluateReturnExpression (alpine.js:1705)
    at handleAttributeBindingDirective (alpine.js:632)
    at alpine.js:1668
    at Array.forEach (<anonymous>)
    at Component.resolveBoundAttributes (alpine.js:1654)
    at Component.updateElement (alpine.js:1630)
    at alpine.js:1588
    at alpine.js:1555
```

See: https://codepen.io/smeedijzer-ijzersterk-online/pen/yLJMzMj

Any idea why and how to fix it?

**Comment by MuzafferDede** [comment, marked as answer]
Seems like you have messed with the Global scope variable names and data scope on `slides` 

instead of binding slides on init, bind it via function as 

```js
rawSlides = [
	{
		image: "https://picsum.photos/seed/picsum/1920/1080",
		quote: "Lorem ipsum dolor sit amet",
		name: "John Doe",
		company: "Company"
	},
	{
		image: "https://picsum.photos/seed/picsum2/1920/1080",
		quote: "Consectetuer adipiscing elit",
		name: "Jane Doe",
		company: "Code Company"
	}
];

sliderData = function (slides) {
	return {
....
                slides: function(){
                   return rawSlides.map((slide, index) => {				
				slide.id = index + Date.now();
				return slide;
			});
                },

		init() {
			this.initSwiper();
		},
```

```html
<div class="h-full" x-data="sliderData()" x-init="init()">
....
<template x-for="(slide, index) in slides()" :key="slide.id">
....
```

Notice function calls vs variable names and where they are used.

**Reply by huubl** [reply]
Thanks for helping.

I changed the code, unfortunately the the error persists. I found that changing the Swiper property: `loop` to `false` removes the errors, but the background image is not loading after the 1st slide.

https://codepen.io/smeedijzer-ijzersterk-online/pen/qBNYraY

I'm not sure if this is due Alpine or SwiperJS

**Reply by MuzafferDede** [reply]
After a quick review, my thoughts: 
I noticed that `template` element might cause the issue since swiper is not handling template tag and it inserts it into the slide list. Loop clones slide elements and `template` element treated as a slide item imo. I have similar use case with Vue and it works normal. Since Alpine rely on template tag for `x-for` to do its thing, maybe swiper and alpine doesn't work well together.

**Reply by huubl** [reply]
That makes sense. Probably Alpine and Swiper are not the way to go.

Thanks for looking in to it!

**Comment by sky-xo** [comment]
Instead of using `x-for`, one solution/hack is to create a static amount of slides (the most you think you'll ever need), and the use `x-show` to hide any slides that don't have content. For instance:

```
<div class="swiper-slide" x-show="getSlide(0)"> ... </div>
<div class="swiper-slide" x-show="getSlide(1)"> ... </div>
<div class="swiper-slide" x-show="getSlide(2)"> ... </div>
```

You just need to create a `getSlide` method in swiperData that returns null or false if the slide doesn't exist.
